zeth-contracts/contracts/Groth16Mixer.sol

```diff
@@ -43,31 +83,274 @@ contract Groth16Mixer is BaseMixer {
         check_mkroot_nullifiers_hsig_append_nullifiers_state(vk, input);
 
         // 2.a Verify the signature on the hash of data_to_be_signed
-        bytes32 hash_to_be_signed = sha256(abi.encodePacked(
-            pk_sender, ciphertext0, ciphertext1, a, b, c, input));
+        bytes32 hash_to_be_signed = sha256(
+            abi.encodePacked(
+                pk_sender,
+                ciphertext0,
+                ciphertext1,
+                a,
+                b,
+                c,
+                input
+            ));
         require(
-            otsig_verifier.verify(vk, sigma, hash_to_be_signed),
+            OTSchnorrVerifier.verify(
+                vk[0], vk[1], vk[2], vk[3], sigma, hash_to_be_signed),
             "Invalid signature: Unable to verify the signature correctly"
         );
 
         // 2.b Verify the proof
         require(
-            zksnark_verifier.verifyTx(a, b, c, input),
+            verifyTx(a, b, c, input),
             "Invalid proof: Unable to verify the proof correctly"
         );
 
         // 3. Append the commitments to the tree
         assemble_commitments_and_append_to_state(input);
 
-        // 4. get the public values in Wei and modify the state depending on
+        // 4. Get the public values in Wei and modify the state depending on
         // their values
         process_public_values(input);
 
         // 5. Add the new root to the list of existing roots and emit it
         add_and_emit_merkle_root(getRoot());
 
-        // Emit the all the coins' secret data encrypted with the recipients'
+        // 6. Emit the all the coins' secret data encrypted with the recipients'
         // respective keys
         emit_ciphertexts(pk_sender, ciphertext0, ciphertext1);
     }
+
+    function verify(
+        uint[] memory input,
+        Proof memory proof) internal returns (uint) {
+
+        // `input.length` = size of the instance = l (see notations in the
+        // reference paper).  We have coefficients indexed in the range[1..l],
+        // where l is the instance size, and we define a_0 = 1. This is the
+        // reason we need to check that: input.length + 1 == vk.ABC.length (the
+        // +1 accounts for a_0). This equality is a strong consistency check
+        // (len(givenInputs) needs to equal expectedInputSize (not less))
+        require(
+            input.length + 1 == verifyKey.ABC.length,
+            "Input length differs from expected");
+
+        // A memory scratch pad
+        uint[24] memory pad;
```

Comment:
Why 24?

Reply:
24 comes from the largest capacity required by the implementation (0x300 bytes).  I've added an explanation just before this decl (trying to avoid explicit constants in the comment, so they don't get invalidated by future changes).